src/posts/cubed-xarray/index.md

```diff
@@ -0,0 +1,208 @@
+
+This type of failure occurs sometimes for real array workloads when using dask, but is a symptom of the general difficulty of memory management when using any cluster-based framework. (The performance of on this particular case could almost certainly be improved by changes to dask's distributed scheduler, and indeed [`dask.distributed` has gotten hugely better at memory management recently](https://medium.com/pangeo/dask-distributed-and-pangeo-better-performance-for-everyone-thanks-to-science-software-63f85310a36b). Here though we are just using this case as a specific example of a type of failure that is very difficult to solve in general.)
+
+### Cubed results
+
+[Trying this problem with Cubed](https://gist.github.com/TomNicholas/c6a28f7c22c6981f75bce280d3e28283) (using the Lithops executor and Google Cloud Functions), we find that Cubed successfully completed the workload at all scales, including the 1.5TB workload. Remarkably, it did so using only 1.3GB of RAM per serverless container, staying below the threshold of 2GB we set in the `cubed.Spec`.
+
+If we take just the largest version of this workload (1.5TB), we can visualize the relationship between the various types of memory Cubed accounts for over the course of the computation.
+
+![](https://hackmd.io/_uploads/rylFHft_n.png)
+
+FIGURE: Cubed's actual memory usage vs projected memory usage, for the largest workload (1.5TB). Cubed successfully processed 1.5TB of data whilst only using a maximum of 1.3GB of RAM per serverless container, staying well below the given limit of 2GB of RAM. This was run using Lithops on Google Cloud Functions, which provisioned thousands of serverless containers each with 2048MB of RAM.
+
+You can see again that the [projected memory](https://tom-e-white.com/cubed/user-guide/memory.html) usage is below the allowed memory usage (else Cubed would have raised an exception before the job even started running), and the actual peak memory used was lower still. We've also plotted the reserved memory, which is a parameter intended to account for the memory usage of the executor itself (i.e. Lithops here), and was estimated by measuring beforehand using [`cubed.measure_reserved_memory()`](https://tom-e-white.com/cubed/generated/cubed.measure_reserved_memory.html#cubed.measure_reserved_memory).
+
+One obvious tradeoff for this memory stability is that Cubed took **a lot longer** to complete - roughly 4x longer then dask for the 1.5TB workload (45m 22s vs 11m 26s). We will come back to discuss this shortly.
+
+Finally it's interesting to look at Cubed's equivalent of the task graph. To calculate one array (the product $<UV>_t$ from the quadratic means problem), Cubed's "Plan" for processing 1.5TB of data looks like this:
```

Comment:
`$<UV>_t$` is problematic... it results in this error

```
SyntaxError: unknown: Expected corresponding JSX closing tag for <UV>.
```

Reply:
Okay, is latex formatting allowed? I just tried changing it to `$UV$` to see if that helps

Reply:
a subset of it is supported. i intend to add the necessary plugins required to support the missing features